Bug 1194830 - Add button to trigger new buildbot jobs

adusca opened the pull request:
This adds a '+' button to resultsets that allows users to trigger new buildbot jobs.

Screenshots of this in action are in: https://drive.google.com/folderview?id=0B7rHgvgC7s4ZflY4cmtSSUJjcHVqQktIMGljcE5nLW9jaWx5T21CQ3F5QllSVGtDQl9sSGM&usp=sharing

When users click on '+' TH renders every possible job. Users can then click on the jobs they want and then click on 'Trigger New Jobs' to trigger the selected jobs. This will send a pulse message that Pulse Actions will use to trigger the jobs.

Currently it takes a long time (~ 7 seconds) for the full list of possible jobs to show up, because it is downloading and processing the list of builders every time. I'm working on adding this info to the TH database by adding a table and running an ETL job to fill it.

I'm planning on dealing with authentication on the Pulse Actions side using the same logic as in Try Extender, that is, only triggering jobs if the user is the commit author or has a '@mozilla.com' address, but if there is a solution on TH side I would be happy to use it.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/892)
<!-- Reviewable:end -->

adusca commented on treeherder/etl/buildbot.py:
```diff
@@ -1066,3 +1067,46 @@ def get_device_or_unknown(job_name, vm):
+def get_all_possible_jobs_for_project(project):
+    """Return a list with all buildbot jobs that can be triggered on a project."""
+    all_builders = requests.get(
+        'http://explique.me/all_builders.json').json()
```
This is a temporary URL, I'm working on getting a more official data source.

tojonmz commented:
So, we have 2 retrigger menu items already in the resultset action menu (when you are logged in and eligible to retrigger). Would it make sense for it to be there?

eg. `Trigger New Buildbot Jobs`


![resultsetactionmenu](https://cloud.githubusercontent.com/assets/3660661/9395650/b0b220a0-475d-11e5-9693-4a5d7937d939.jpg)

adusca commented:
The new commit adds processing allthethings.json to the ETL layer. In order to do that I had to create a new 'possible_jobs' table and a new 'fetch_allthethings' task to celery. I still need to add a proper api endpoint to retrieve the list, but I'm going to leave that for tomorrow :)

maurodoglio commented on treeherder/model/sql/template_schema/project.sql.tmpl:
```diff
@@ -153,6 +153,33 @@ CREATE TABLE `job` (
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
 /*!40101 SET character_set_client = @saved_cs_client */;
 
+--
```
you should either put this table in the reference data schema or remove the repository column as it is redundant

maurodoglio commented on treeherder/model/derived/jobs.py:
```diff
@@ -1250,6 +1257,81 @@ def store_job_data(self, data, raise_errors=False):
                 executemany=True)
 
+    def store_possible_job_data(self, data, raise_errors=False):
+        """
```
I think you need a strategy to modify the possible jobs when the list changes upstream

adusca replied:
My plan was to delete the table every day and create a new one from scratch with the latest allthethings.json, but thinking more about it this sounds wasteful. If there was a way of getting just the daily diff and add/remove rows from the database accordingly it would be much better. I'll try to come up with something.

maurodoglio commented on treeherder/webapp/api/possible_jobs.py:
```diff
@@ -0,0 +1,30 @@
+    This viewset is responsible for the possible_jobs endpoint.
+
+    """
+    throttle_scope = 'jobs'
```
I would either use a custom throttle_scope or none.

camd commented:
You might add a ``thNotify.send`` message if the user clicks the + icon, but there are not jobs to add.  Just in case either the table is not populated, or all jobs have been run already.

camd commented:
We need a way to turn the "possible jobs" back off.  Perhaps the + button should be a toggle where clicking a second time will turn them off.

camd commented:
FWIW: I agree with @tojonmz about adding this to the action menu, rather than a button on the toolbar.  Perhaps wording like "Select backfill jobs" or something like that.  Sometimes an ellipsis after the text indicates there is more action to take.  Though often that means a dialog will pop up.  So not sure if that applies here.  Maybe @tojonmz has an opinion?

tojonmz commented:
> You might add a thNotify.send message if the user clicks the + icon, but there are not jobs to add. Just in case either the table is not populated, or all jobs have been run already.

Or, if there is a way of knowing there are no jobs to populate, apply a disabled class to the + button in the resultset bar with a suitable `title="No jobs to add"`.

camd commented on treeherder/client/thclient/client.py:
```diff
@@ -349,6 +349,17 @@ def init_data(self):
             }
 
 
+class TreeherderPossibleJob(TreeherderData, ValidatorMixin):
```
Thinking about this more, I'm not sure we need to enable this in the ``thclient``.  This is really just an internal process.  We aren't going to have external folks post this data to TH, are we?  If so, then cool as it is.  But if not, I think we could just have the celery task write directly to the DB (through ``jobs.py`` as you're doing) without going through a call to an API.

tojonmz commented:
> Sometimes an ellipsis after the text indicates there is more action to take. Though often that means a dialog will pop up.

Correct, the use of "Menu entry..." implies a dialog of some kind is launched.

camd commented on ui/index.html:
```diff
@@ -200,6 +200,15 @@
                   title="{{ title }}">{{ value }}</button>
         </script>
 
+        <!-- Possible Job Btn span -->
```
as mentioned earlier, I'd vote for adding this to the action list instead of a button here.  :)

tojonmz replied:
+1

adusca commented:
Thanks for the reviews! I am almost done addressing the comments, but not quite. The only thing missing is hiding the possible job list appropriately. In particular, I am using @camd's suggestion of inserting data directly from the task instead of going through an endpoint. I also moved the button to the dropdown menu

maurodoglio commented:
@adusca is this ready for a final review?

tojonmz commented:
I'm going to try to test this PR locally today. As mentioned in channel I'm concerned we aren't providing consistent selection workflow to all jobs, now that you can select multiple 'possible' jobs in this PR (which is great) but you can't select multiple jobs of any other kind.

tojonmz commented:
I did a quick first pass, but I wasn't able to get any farther than just launching the branch and loading m-c, m-i, and try, and selecting the new Action menu entry. Any resultsets I selected the "Add new Jobs" action menu with, didn't yield any new jobs in the jobs table, for additional testing.

Unrelated, and I know this is late in the game to raise it (I wasn't aware of a new job type until now) but the internal use of 'possible'; I wonder if we have words with better semantics? To me 'possible job' means... "I might be a job or I might not be a job". Some other candidates throughout the PR:

"launchable_job"
"new_job" (since we use it in the menu)
"runnable_job" (ties in nicely with our concept of 'running') - I like this one of the three posed.

With corresponding changes everywhere in the PR to apply that naming convention.

---

Review status: 0 of 25 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

<sup>**[ui/js/controllers/jobs.js, line 175 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9FKJ6mOok4eVqUzhQ)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/controllers/jobs.js#L175)):</sup>
This `$rootScope.$emit` could be consolidated on one line, I think.

---

<sup>**[ui/js/controllers/jobs.js, line 233 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9Fs6Ypcqa_R1kGIwr)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/controllers/jobs.js#L233)):</sup>
Just a nit, we could add a space between the `()` and the `{` here for niceness. I know we have it scattered throughout the repo.

The same we could do elsewhere in the PR.

---

<sup>**[ui/js/controllers/jobs.js, line 238 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9GQVhPpBTNtIB2TYp)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/controllers/jobs.js#L238)):</sup>
Same here.

---

<sup>**[ui/js/directives/treeherder/clonejobs.js, line 247 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9GoBRnirS-9m3BRje)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/directives/treeherder/clonejobs.js#L247)):</sup>
Not sure, but I think these two lines above can be lined up under the `$`.

---

<sup>**[ui/js/directives/treeherder/clonejobs.js, line 857 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9H4dqQGSCC6t96NoL)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/directives/treeherder/clonejobs.js#L857)):</sup>
This could probably be moved up a line to save an indent. And same with `() {` here.

---

<sup>**[ui/js/models/job.js, line 28 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9Hif6iDT1xdUT62yP)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/models/job.js#L28)):</sup>
We could break this function out onto multiple lines if you felt like it.

---

<sup>**[ui/partials/main/thActionButton.html, line 16 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9AHggzNq2KQZ891Z8)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/partials/main/thActionButton.html#L16)):</sup>
Per my comment in the PR thread [here](https://github.com/mozilla/treeherder/pull/892#issuecomment-134695365) I think we should either:

1) conditionally apply a disabled class to this menu and a title of "No new jobs available" for resultsets whose `possible_jobs/` endpoints have a count/length of 0.

2) or if we aren't able to do that, then we should `thNotify` with the blue "info" style, the equivalent message: "No new jobs available" when the function is called.

Otherwise the user selects the menu and doesn't know if something happened or not, or if something is happening with a delay but they can't see it. I just experienced this in local testing :)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/892)
<!-- Sent from Reviewable.io -->

tojonmz commented:
Ok, I managed to get 'new jobs' appearing in the UI this morning in my local branch.

The job containers for these new jobs need to be adjusted, they are being drawn with a border in all cases, and the dimensions are too large, they are altering the row height. They should be identical in computed dimension to all other jobs.

In the current PR in my opinion the best approach would be simply to:

* implement a job style that uses an exact copy of the pending class - and instead just add a new property `text-decoration: overline;` to this new job type (since pending and this new type are so analogous)

or

* use a copy of pending and and simply give it some alternate color (+)

And:

* ensure the new job type is represented correctly in (+n) job counts with the same style
* ensure the new job type is the same container dimension as all other jobs

(+) If you went with color we just need to ensure whatever new color gets used, is complimentary to our existing jobs and doesn't get confused with any other job result-state. It's becoming more problematic as we have many job types, but the current used colors are here.
https://treeherder.mozilla.org/userguide.html

**Unrelated:** How do I turn off "new" jobs once I've made them visible?

**Unrelated:** Based on being able to see the jobs now, I would favour "runnable" as the chosen descriptor used throughout this PR (per my earlier comment). It makes the title on the job hover make more sense than 'possible' imo.

---

Review status: 0 of 25 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/892)
<!-- Sent from Reviewable.io -->

tojonmz commented:
@adusca @camd I did a bunch of possible job icon treatments for the new 'runnable' and a revised pending; I'll share them here shortly.

I was also wondering, if there's difficulty in turning 'runnable' jobs off once they've been made visible by this new function, maybe instead we could load them at all times - but filter them off by default? Users could expose them via the global Filter panel instead, as a new job type (which it is anyway).

tojonmz commented:
Not to hijack the PR, but here's my proposal for the new *'runnable'*, and pending, and running jobs.

Adding a few folks for visibility, @rvandermeulen @KWierso @camd @edmorley @jonallengriffin 

Rationale:
* we use an adjacent symbol (star*) for classified jobs - why not use this concept for other states?
* leverage that concept for the new 'runnable', and also pending jobs
* these states are always transient so exert no permanent UI distraction
* this is partly inspired by Travis CI state colors (amber, green) during a run

I did about 10-12, these ended up being my favorites so far. The green triangle is actually bootstrap 'yellowgreen'. Both it and the open square hold up well at normal size in Treeherder on my monitor. Runnable and pending are so similar in concept, I pose they should share the same grey and be differentiated by the symbol (and its color for those not colorblind).

Mockups in each row below:
* running is the first job, and is unchanged
* pending is the second job in lightgrey
* runnable is the third job also in lightgrey

My favorite:
![fa-square-o_runnable](https://cloud.githubusercontent.com/assets/3660661/9702642/5504beda-5436-11e5-90cc-8f4b8cad3315.jpg)
Another alternate but fa-spinner didn't hold up as well:
![fa-spinner-pending](https://cloud.githubusercontent.com/assets/3660661/9702644/5e9988a4-5436-11e5-849f-305221cdb572.jpg)
Other variation with fa-poweroff for runnable, but didn't hold up as well:
![fa-poweroff_runnable](https://cloud.githubusercontent.com/assets/3660661/9702646/697662a6-5436-11e5-81c7-9068b5370ff9.jpg)
Other variation but circle wasn't as strong as square - which is more evocative of 'stopped'
![fa-circle-o_runnable](https://cloud.githubusercontent.com/assets/3660661/9702647/6cb69486-5436-11e5-8a83-d86b9c04a7b0.jpg)

This treatment should also work fine with (+n) job counts. If we're interested in pursing this, let me know and I will split this off into a separate bug/PR that we can land at the same time as this one :)

adusca commented:
I renamed possible jobs to runnable jobs; also, it is now possible to hide runnable jobs by clicking on "Hide Runnable Jobs", and the jobs are also automatically hidden when you click on "Trigger New Jobs".

I think I addressed every review comment. I would be happy to coordinate landing this with the css changes proposed by @tojonmz, or if you would like me to change the css in this patch to match the proposed one I would also be happy.

I asked :mdoglio for a new round of reviews on bugzilla, is this the appropriate time to ask for a new round from @camd and @tojonmz or should I wait for the results of the first review?

maurodoglio commented:
Reviewed 3 of 16 files at r1, 2 of 16 files at r2, 2 of 18 files at r4, 20 of 20 files at r5.
Review status: all files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

<sup>**[treeherder/etl/allthethings.py, line 38 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JymEu1uMefl1zYng-bn)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/treeherder/etl/allthethings.py#L38)):</sup>
Cleaning up the table in this way is not safe as something could go wrong between the deletion of the old content and the insertion of the new one. You should instead:
- Remove the content not present in the latest allthethings file.
- Insert only new rows.
- Update already existing content with the latest from the allthethings file

---

<sup>**[treeherder/model/derived/jobs.py, line 1246 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JymPc2u_hXqxlbqx_hH)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/treeherder/model/derived/jobs.py#L1246)):</sup>
I'm wondering if we can store the allthethings file in a denormalized way, making it completely independent of the rest. That would imply storing all the data as varchar instead of foreign keys. Since the dataset is capped to a maximum size (the number of possible buildernames) that wouldn't be too bad. @emorley what do you think?

---

<sup>**[treeherder/model/sql/template_schema/project.sql.tmpl, line 153 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxD_qTiKvQLZmerWSLR-r2-156)** ([raw file](https://github.com/mozilla/treeherder/blob/3467f2e792de071d6ab6dfe37fd95e2d0733010a/treeherder/model/sql/template_schema/project.sql.tmpl#L153)):</sup>
We are trying to move away from using JobsModel and use the django orm insted. It's probably worth to put this table directly in the reference schema

---

<sup>**[treeherder/model/tasks.py, line 109 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JymY9ZdYZznf9LyTAHf)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/treeherder/model/tasks.py#L109)):</sup>
Can you please define a different task to publish this kind of action? This way you can make the list of buildernames a required parameter. On the long term the solution is to fix https://bugzilla.mozilla.org/show_bug.cgi?id=1197197

---

<sup>**[treeherder/settings/base.py, line 296 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JymJXUT54Lpwws-_d2B)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/treeherder/settings/base.py#L296)):</sup>
We usually don't set a throttling policy on GET methods, you can remove this.

---

<sup>**[ui/js/controllers/jobs.js, line 230 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JymfiKw4VgVwDbujKnQ)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/ui/js/controllers/jobs.js#L230)):</sup>
If the show/hide runnable jobs is part of the resultset menu there should be no need for this function. You can instead directly use $scope.resultset.isRunnableVisible

---

<sup>**[ui/js/models/job.js, line 36 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jyme9pt34DaPu_xTq5r)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/ui/js/models/job.js#L36)):</sup>
I don't think it's a good idea to put a job dress on something that is not a job. Can you please create a new ThRunnableJobModel class instead?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/892)
<!-- Sent from Reviewable.io -->

adusca commented:
Review status: 5 of 29 files reviewed at latest revision, 16 unresolved discussions, some commit checks broke.

---

<sup>**[treeherder/etl/allthethings.py, line 0 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JymEu1uMefl1zYng-bn)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/treeherder/etl/allthethings.py#L0)):</sup>
Done in a different way: I created a 'last_touched' datetime field for each runnable job. This field gets updated whenever we see the buildername in allthethings.json, and the endpoint only returns builders that have been touched in the last week.
<!-- Reviewable comment -K-L5f4YLEL01DZsANCL:-428753552 -->

---

<sup>**[treeherder/model/tasks.py, line 0 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JymY9ZdYZznf9LyTAHf)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/treeherder/model/tasks.py#L0)):</sup>
Done.
<!-- Reviewable comment -K-L6Q1N7G-8iXWcc2WY:-499146009 -->

---

<sup>**[treeherder/settings/base.py, line 0 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JymJXUT54Lpwws-_d2B)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/treeherder/settings/base.py#L0)):</sup>
Done.
<!-- Reviewable comment -K-L6QipCOBouuBaIB_f:-499146009 -->

---

<sup>**[ui/index.html, line 0 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxBoHiu27igf9q7e9cD-r3-203)** ([raw file](https://github.com/mozilla/treeherder/blob/96499699eca11a9664d549500140b794a83d9863/ui/index.html#L0)):</sup>
Done.
<!-- Reviewable comment -K-L6RqjZp97iVZCpkVl:-499146009 -->

---

<sup>**[ui/js/controllers/jobs.js, line 0 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9FKJ6mOok4eVqUzhQ)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/controllers/jobs.js#L0)):</sup>
Done.
<!-- Reviewable comment -K-L6SZbUznP8PEuL6yR:-499146009 -->

---

<sup>**[ui/js/controllers/jobs.js, line 0 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9Fs6Ypcqa_R1kGIwr)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/controllers/jobs.js#L0)):</sup>
Done.
<!-- Reviewable comment -K-L6SxoknlUNNyWy5dJ:-499146009 -->

---

<sup>**[ui/js/controllers/jobs.js, line 0 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9GQVhPpBTNtIB2TYp)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/controllers/jobs.js#L0)):</sup>
Done.
<!-- Reviewable comment -K-L6TEOutBiOk-3CWXo:-499146009 -->

---

<sup>**[ui/js/controllers/jobs.js, line 0 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JymfiKw4VgVwDbujKnQ)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/ui/js/controllers/jobs.js#L0)):</sup>
Done.
<!-- Reviewable comment -K-L6UChPn4pIRXJvtGG:-499146009 -->

---

<sup>**[ui/js/directives/treeherder/clonejobs.js, line 0 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9GoBRnirS-9m3BRje)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/directives/treeherder/clonejobs.js#L0)):</sup>
Done.
<!-- Reviewable comment -K-L6VDkV5eF-m8NKWxw:-499146009 -->

---

<sup>**[ui/js/directives/treeherder/clonejobs.js, line 0 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9H4dqQGSCC6t96NoL)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/directives/treeherder/clonejobs.js#L0)):</sup>
Done.
<!-- Reviewable comment -K-L6VXO2vEQnGrBem31:-499146009 -->

---

<sup>**[ui/js/models/job.js, line 28 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9Hif6iDT1xdUT62yP)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/js/models/job.js#L28)):</sup>
Done.
<!-- Reviewable comment -K-L6Y6gFtpN9KE_VDEx:-499146009 -->

---

<sup>**[ui/js/models/job.js, line 36 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jyme9pt34DaPu_xTq5r)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/ui/js/models/job.js#L36)):</sup>
I created a new (minimal) class here. Can you clarify the scope of the wanted change?
<!-- Reviewable comment -K-L8oa_cVjBAQvYsjB2:953619103 -->

---

<sup>**[ui/partials/main/thActionButton.html, line 0 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-Jy9AHggzNq2KQZ891Z8)** ([raw file](https://github.com/mozilla/treeherder/blob/82dc1f32d4673cd45dc0c9993407a35301d6b4c0/ui/partials/main/thActionButton.html#L0)):</sup>
As of a few patches ago, we do option 2.
<!-- Reviewable comment -K-L6dDFuh0dTZ_XtfiY:-1035786600 -->

---

<sup>**[treeherder/client/thclient/client.py, line 0 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxD_qTiKvQLZmerWSLM-r3-352)** ([raw file](https://github.com/mozilla/treeherder/blob/96499699eca11a9664d549500140b794a83d9863/treeherder/client/thclient/client.py#L0)):</sup>
Done.
<!-- Reviewable comment -K-L6jqTmYGQXPfEDWUm:-499146009 -->

---

<sup>**[treeherder/model/derived/jobs.py, line 0 \[r5\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JymPc2u_hXqxlbqx_hH)** ([raw file](https://github.com/mozilla/treeherder/blob/31443e4ca830ea96dd84b14a5e28841099084a03/treeherder/model/derived/jobs.py#L0)):</sup>
I only did this for the reference_data_signatures table reference, but it would be easy to completely denormalize the table. @maurodoglio and @emorley, what should I do?
<!-- Reviewable comment -K-L69Z6Ejv7oPNQGz9C:244490510 -->

---

<sup>**[treeherder/model/sql/template_schema/project.sql.tmpl, line 0 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxD_qTiKvQLZmerWSLR-r2-156)** ([raw file](https://github.com/mozilla/treeherder/blob/3467f2e792de071d6ab6dfe37fd95e2d0733010a/treeherder/model/sql/template_schema/project.sql.tmpl#L0)):</sup>
Done.
<!-- Reviewable comment -K-L6OttTNTw3xLJLEku:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/892)
<!-- Sent from Reviewable.io -->

camd commented on ui/js/directives/treeherder/clonejobs.js:
```diff
@@ -196,10 +200,24 @@ treeherder.directive('thCloneJobs', [
+        var filterWithRunnable = function(job) {
+            var rsMap = ThResultSetStore.getResultSetsMap($rootScope.repoName);
+            var visible = thJobFilters.showJob(job);
+            if (job.state === "runnable")
```
Need ``{`` and ``}`` braces around the action clause

camd commented on ui/js/directives/treeherder/clonejobs.js:
```diff
@@ -196,10 +200,24 @@ treeherder.directive('thCloneJobs', [
         };
 
+        var filterWithRunnable = function(job) {
+            var rsMap = ThResultSetStore.getResultSetsMap($rootScope.repoName);
```
I would move this get of ``rsMap`` to inside the ``if`` statement, so you don't waste time assigning if the job isn't ``runnable``.

camd commented on ui/js/directives/treeherder/clonejobs.js:
```diff
@@ -230,7 +248,19 @@ treeherder.directive('thCloneJobs', [
+                if (ThResultSetStore.isRunnableJobSelected($rootScope.repoName,
+                                                           job.result_set_id,
+                                                           jobStatus.buildername))
+                    jobBtn.addClass("runnable-job-btn-selected");
```
also need { } around this line

camd commented on ui/js/directives/treeherder/clonejobs.js:
```diff
@@ -230,7 +248,19 @@ treeherder.directive('thCloneJobs', [
+                                                           jobStatus.buildername))
+                    jobBtn.addClass("runnable-job-btn-selected");
+            }
+            else{
```
space before the {.

Sorry for all the nits, but we try to keep (are trying to get) our style consistent.  :)

camd commented on ui/js/directives/treeherder/clonejobs.js:
```diff
@@ -356,8 +386,11 @@ treeherder.directive('thCloneJobs', [
-            if(key && !_.isEmpty(this.job_map[key])){
+            if (buildername) {
+                _.bind(clickRunnableJobCb, this, el, resultset.id)();
+            } else if(key && !_.isEmpty(this.job_map[key])){
```
should be:
    } else if (key && !_.isEmpty(this.job_map[key])) {

camd commented on ui/js/models/resultsets_store.js:
```diff
@@ -272,6 +274,28 @@ treeherder.factory('ThResultSetStore', [
             return key;
         };
 
+        var getRunnableJobs = function(repoName, resultSet) {
```
this might be better named as ``addRunnableJobs`` since it doesn't return the jobs themselves, but takes the action of adding them into the joblist.

camd commented on ui/js/models/resultsets_store.js:
```diff
@@ -786,6 +810,29 @@ treeherder.factory('ThResultSetStore', [
         };
 
+        var getSelectedRunnableJobs = function(repoName, resultsetId){
+            if (!repositories[repoName].rsMap[resultsetId].selected_runnable_jobs)
```
nit: enclose in { }

camd commented on ui/js/models/resultsets_store.js:
```diff
@@ -786,6 +810,29 @@ treeherder.factory('ThResultSetStore', [
+        };
+
+        var isRunnableJobSelected = function(repoName, resultsetId, buildername){
+            var selectedRunnableJobs = getSelectedRunnableJobs(repoName, resultsetId);
```
not a big deal, but you can use this function for stuff like this: https://lodash.com/docs#includes

edmorley commented:
Had a quick first pass (on a train at the moment). Will take a better look over the weekend - in the meantime if you could rebase onto master and add the new migrations file, that would be great 
<!-- Reviewable comment -K31M5Dx7Mi8X5pNoQrX:1292042081 -->

---

Reviewed 1 of 30 files at r6, 1 of 20 files at r7, 1 of 18 files at r8.
Review status: 4 of 28 files reviewed at latest revision, 27 unresolved discussions.

---

<sup>**[treeherder/etl/allthethings.py, line 111 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-K31KKRE7_39dK41FdZB)** ([raw file](https://github.com/mozilla/treeherder/blob/6c2dc690d06692dd36d8941a14987a849b660646/treeherder/etl/allthethings.py#L111)):</sup>
I don't think we need to return True here - nothing uses the return value & we never return False. (If there was an error, we'd raise an exception instead).
<!-- Reviewable comment -K31KKRFFYwseqRaHRF6:-752158111 -->

---

<sup>**[treeherder/etl/tasks/buildapi_tasks.py, line 38 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-K31KZmYB1BOy2myyngL)** ([raw file](https://github.com/mozilla/treeherder/blob/6c2dc690d06692dd36d8941a14987a849b660646/treeherder/etl/tasks/buildapi_tasks.py#L38)):</sup>
How long does the task typically take? Would something less than 10 mins be enough?
<!-- Reviewable comment -K31KZmZZn9zIjWu8Ne5:1697390008 -->

---

<sup>**[treeherder/model/migrations/0001_initial.py, line 300 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-K31Kp4QKOB5jeb6vGW8)** ([raw file](https://github.com/mozilla/treeherder/blob/6c2dc690d06692dd36d8941a14987a849b660646/treeherder/model/migrations/0001_initial.py#L300)):</sup>
We're now using Django migrations, so the existing files in this directory should not be modified. Instead, run `./manage.py makemigrations --name add_runnable_job_table` from the vagrant environment, to create a new migrations file automatically.
<!-- Reviewable comment -K31Kp4QKOB5jeb6vGW9:1674231867 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/892)
<!-- Sent from Reviewable.io -->

adusca commented:
Review status: 4 of 27 files reviewed at latest revision, 27 unresolved discussions.

---

<sup>**[treeherder/etl/allthethings.py, line 111 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-K31KKRE7_39dK41FdZB)** ([raw file](https://github.com/mozilla/treeherder/blob/6c2dc690d06692dd36d8941a14987a849b660646/treeherder/etl/allthethings.py#L111)):</sup>
Done.
<!-- Reviewable comment -K3BJYgJbdgGJzeTBHDh:-499146009 -->

---

<sup>**[treeherder/etl/tasks/buildapi_tasks.py, line 38 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-K31KZmYB1BOy2myyngL)** ([raw file](https://github.com/mozilla/treeherder/blob/6c2dc690d06692dd36d8941a14987a849b660646/treeherder/etl/tasks/buildapi_tasks.py#L38)):</sup>
I increased the time limit from 3 minutes to 10 because I was getting intermittents time outs. Testing now with a more powerful computer the task took 1min32s.
<!-- Reviewable comment -K3BIfAaSxiROb3gkWN0:-1502194506 -->

---

<sup>**[ui/js/directives/treeherder/clonejobs.js, line 214 \[r7\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxBoHiu27igf9q7e9cF-r7-214)** ([raw file](https://github.com/mozilla/treeherder/blob/82711e5e6f048ff579add1a16a86b02db27d82cb/ui/js/directives/treeherder/clonejobs.js#L214)):</sup>
Done.
<!-- Reviewable comment -K3BJFnzexbtzIYbOCor:-499146009 -->

---

<sup>**[ui/js/directives/treeherder/clonejobs.js, line 216 \[r7\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxBoHiu27igf9q7e9cF-r7-216)** ([raw file](https://github.com/mozilla/treeherder/blob/82711e5e6f048ff579add1a16a86b02db27d82cb/ui/js/directives/treeherder/clonejobs.js#L216)):</sup>
Done.
<!-- Reviewable comment -K3BJdSQjztcXZKfozIH:-499146009 -->

---

<sup>**[ui/js/directives/treeherder/clonejobs.js, line 259 \[r7\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxBoHiu27igf9q7e9cF-r7-259)** ([raw file](https://github.com/mozilla/treeherder/blob/82711e5e6f048ff579add1a16a86b02db27d82cb/ui/js/directives/treeherder/clonejobs.js#L259)):</sup>
Done.
<!-- Reviewable comment -K3BJJWYCt6Cj2BcxGVZ:-499146009 -->

---

<sup>**[ui/js/directives/treeherder/clonejobs.js, line 261 \[r7\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxBoHiu27igf9q7e9cF-r7-261)** ([raw file](https://github.com/mozilla/treeherder/blob/82711e5e6f048ff579add1a16a86b02db27d82cb/ui/js/directives/treeherder/clonejobs.js#L261)):</sup>
Done.
<!-- Reviewable comment -K3BJK1_K64mUOdYJI-7:-499146009 -->

---

<sup>**[ui/js/directives/treeherder/clonejobs.js, line 393 \[r7\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxBoHiu27igf9q7e9cF-r7-393)** ([raw file](https://github.com/mozilla/treeherder/blob/82711e5e6f048ff579add1a16a86b02db27d82cb/ui/js/directives/treeherder/clonejobs.js#L393)):</sup>
Done.
<!-- Reviewable comment -K3BJV9nTtrWunCFOOkK:-499146009 -->

---

<sup>**[ui/js/models/resultsets_store.js, line 277 \[r7\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxBoHiu27igf9q7e9cI-r7-277)** ([raw file](https://github.com/mozilla/treeherder/blob/82711e5e6f048ff579add1a16a86b02db27d82cb/ui/js/models/resultsets_store.js#L277)):</sup>
Done.
<!-- Reviewable comment -K3BJQWGdc67LOZxYTH1:-499146009 -->

---

<sup>**[ui/js/models/resultsets_store.js, line 814 \[r7\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxBoHiu27igf9q7e9cI-r7-814)** ([raw file](https://github.com/mozilla/treeherder/blob/82711e5e6f048ff579add1a16a86b02db27d82cb/ui/js/models/resultsets_store.js#L814)):</sup>
Done.
<!-- Reviewable comment -K3BJQvXmx6L41VG1vk6:-499146009 -->

---

<sup>**[ui/js/models/resultsets_store.js, line 831 \[r7\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-JxBoHiu27igf9q7e9cI-r7-831)** ([raw file](https://github.com/mozilla/treeherder/blob/82711e5e6f048ff579add1a16a86b02db27d82cb/ui/js/models/resultsets_store.js#L831)):</sup>
Done.
<!-- Reviewable comment -K3BJS9IUCZd0RMOZGFd:-499146009 -->

---

<sup>**[treeherder/model/migrations/0001_initial.py, line 300 \[r8\]](https://reviewable.io:443/reviews/mozilla/treeherder/892#-K31Kp4QKOB5jeb6vGW8)** ([raw file](https://github.com/mozilla/treeherder/blob/6c2dc690d06692dd36d8941a14987a849b660646/treeherder/model/migrations/0001_initial.py#L300)):</sup>
And I was writing by hand :P ! I ran the command and the generated file is included in the commit.
<!-- Reviewable comment -K3BJgob8zs7nBDUg4XY:-1924953991 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/892)
<!-- Sent from Reviewable.io -->

edmorley commented:
Reviewed 1 of 19 files at r4, 4 of 30 files at r6, 6 of 20 files at r7, 8 of 18 files at r8, 11 of 12 files at r9.
Review status: all files reviewed at latest revision, 24 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/892)
<!-- Sent from Reviewable.io -->
